File: apps/desktop/src/lib/plugin.test.ts
import { describe, it, expect } from 'vitest'
import { isValidPluginId, isValidPluginVersion, sanitizePluginPath, validatePluginManifest, validatePluginLoadPayload, canPluginAccess, canTransitionPlugin, mockTranslate, ALLOWED_CAPS } from './plugin'

describe('plugin validation', () => {
  it('valid pluginId', () => {
    expect(isValidPluginId('example-translate')).toBe(true)
    expect(isValidPluginId('ab')).toBe(false)
    expect(isValidPluginId('BadCaps')).toBe(false)
    expect(isValidPluginId('a'.repeat(33))).toBe(false)
  })
  it('valid version', () => {
    expect(isValidPluginVersion('0.1.0')).toBe(true)
    expect(isValidPluginVersion('1.2.3')).toBe(true)
    expect(isValidPluginVersion('0.1')).toBe(false)
    expect(isValidPluginVersion('01.0.0')).toBe(false)
  })
  it('sanitize plugin path', () => {
    expect(sanitizePluginPath('index.js')).toBe(null)
    expect(sanitizePluginPath('src/main.wasm')).toBe(null)
    expect(sanitizePluginPath('../escape.js')).not.toBe(null)
    expect(sanitizePluginPath('/abs.js')).not.toBe(null)
    expect(sanitizePluginPath('evil.txt')).not.toBe(null)
  })
  it('validate manifest ok', () => {
    expect(validatePluginManifest({name:'example-translate', version:'0.1.0', entry:'index.js', capabilities:['notify','clipboard']})).toBe(null)
  })
  it('validate manifest bad caps', () => {
    expect(validatePluginManifest({name:'bad', version:'0.1.0', entry:'index.js', capabilities:['evil']})).not.toBe(null)
  })
  it('validate manifest traversal', () => {
    expect(validatePluginManifest({name:'bad', version:'0.1.0', entry:'../../etc/passwd', capabilities:['notify']})).not.toBe(null)
  })
  it('validate load payload', () => {
    expect(validatePluginLoadPayload({pluginId:'example-translate'})).toBe(null)
    expect(validatePluginLoadPayload({pluginId:'bad/cap'})).not.toBe(null)
    expect(validatePluginLoadPayload({})).not.toBe(null)
  })
  it('can access', () => {
    expect(canPluginAccess(['notify','clipboard'],'notify')).toBe(true)
    expect(canPluginAccess(['notify'],'storage')).toBe(false)
  })
  it('state transitions', () => {
    expect(canTransitionPlugin('UNLOADED','LOADING')).toBe(true)
    expect(canTransitionPlugin('LOADING','LOADED')).toBe(true)
    expect(canTransitionPlugin('LOADED','RUNNING')).toBe(true)
    expect(canTransitionPlugin('RUNNING','RELOADING')).toBe(true)
    expect(canTransitionPlugin('RELOADING','RUNNING')).toBe(true)
    expect(canTransitionPlugin('UNLOADED','RUNNING')).toBe(false)
    expect(canTransitionPlugin('RUNNING','LOADED')).toBe(false)
  })
  it('mock translate', () => {
    expect(mockTranslate('hello','en')).toBe('olleh [translated:en]')
    expect(mockTranslate('Bonjour')).toContain('[translated')
  })
  it('allowed caps', () => {
    expect(ALLOWED_CAPS).toContain('notify')
    expect(ALLOWED_CAPS).toContain('clipboard')
    expect(ALLOWED_CAPS).toContain('storage')
  })
})

File: crates/bridge-core/tests/plugin_test.rs
use bridge_core::{PluginState, validate_plugin_manifest, validate_plugin_load_payload, can_plugin_access, is_valid_plugin_id, is_valid_plugin_version, sanitize_plugin_path, ALLOWED_PLUGIN_CAPS};
use serde_json::json;

#[test]
fn plugin_state_valid() {
    assert!(PluginState::Unloaded.can_transition(&PluginState::Loading));
    assert!(PluginState::Loading.can_transition(&PluginState::Loaded));
    assert!(PluginState::Loaded.can_transition(&PluginState::Running));
    assert!(PluginState::Running.can_transition(&PluginState::Reloading));
    assert!(PluginState::Reloading.can_transition(&PluginState::Running));
    assert!(PluginState::Running.can_transition(&PluginState::Disabled));
}
#[test]
fn plugin_state_invalid() {
    assert!(!PluginState::Unloaded.can_transition(&PluginState::Running));
    assert!(!PluginState::Running.can_transition(&PluginState::Loaded));
}

#[test]
fn is_valid_plugin_id_ok() {
    assert!(is_valid_plugin_id("example-translate"));
    assert!(is_valid_plugin_id("ab-c"));
    assert!(!is_valid_plugin_id("ab"));
    assert!(!is_valid_plugin_id("BadCaps"));
    assert!(!is_valid_plugin_id(&"a".repeat(33)));
}
#[test]
fn is_valid_plugin_version_ok() {
    assert!(is_valid_plugin_version("0.1.0"));
    assert!(is_valid_plugin_version("1.2.3"));
    assert!(!is_valid_plugin_version("0.1"));
    assert!(!is_valid_plugin_version("01.0.0"));
}
#[test]
fn sanitize_plugin_path_ok() {
    assert!(sanitize_plugin_path("index.js").is_ok());
    assert!(sanitize_plugin_path("src/main.wasm").is_ok());
    assert!(sanitize_plugin_path("../escape.js").is_err());
    assert!(sanitize_plugin_path("/abs.js").is_err());
    assert!(sanitize_plugin_path("evil.txt").is_err());
}
#[test]
fn validate_manifest_ok() {
    let m = json!({"name":"example-translate","version":"0.1.0","entry":"index.js","capabilities":["notify","clipboard"],"bridgeVersion":"1"});
    assert!(validate_plugin_manifest(&m).is_ok());
}
#[test]
fn validate_manifest_bad_caps() {
    let m = json!({"name":"bad","version":"0.1.0","entry":"index.js","capabilities":["evil"]});
    assert!(validate_plugin_manifest(&m).is_err());
}
#[test]
fn validate_manifest_traversal() {
    let m = json!({"name":"bad","version":"0.1.0","entry":"../../etc/passwd","capabilities":["notify"]});
    assert!(validate_plugin_manifest(&m).is_err());
}
#[test]
fn validate_load_ok() {
    let p = json!({"pluginId":"example-translate"});
    assert!(validate_plugin_load_payload(&p).is_ok());
}
#[test]
fn validate_load_bad() {
    let p = json!({"pluginId":"bad/cap"});
    assert!(validate_plugin_load_payload(&p).is_err());
}
#[test]
fn can_access() {
    let caps = vec!["notify".to_string(), "clipboard".to_string()];
    assert!(can_plugin_access(&caps, "notify"));
    assert!(!can_plugin_access(&caps, "storage"));
}
#[test]
fn allowed_caps_contains() {
    assert!(ALLOWED_PLUGIN_CAPS.contains(&"notify"));
    assert!(ALLOWED_PLUGIN_CAPS.contains(&"clipboard"));
    assert!(ALLOWED_PLUGIN_CAPS.contains(&"storage"));
    assert!(ALLOWED_PLUGIN_CAPS.contains(&"ai.summarize"));
}
